File: flutter/packages/hatake_core/lib/src/definition/wizard_step_definition.dart
import 'package:equatable/equatable.dart';

import 'field_definition.dart';
import 'form_definition.dart';
import 'layout_definition.dart';
import 'section_definition.dart';

/// One step of a `wizard` page — a [SectionDefinition] with an [id] and a
/// heading. Only this step's [fields] are checked when the user advances.
class WizardStepDefinition extends Equatable {
  /// Stable step identifier.
  final String id;

  /// Step heading.
  final String title;

  /// Optional explanatory text shown under the heading.
  final String? description;

  /// Arrangement of this step's fields.
  final LayoutDefinition layout;

  /// Input fields belonging to this step.
  final List<FieldDefinition> fields;

  /// Show this whole step only when the condition matches what has been entered
  /// so far (see `evaluateCondition`). Null = always shown.
  ///
  /// A hidden step is skipped by 次へ / 戻る **and** is not validated — the step
  /// becomes a section ([form]), so the "a hidden section is not validated" rule
  /// does the work. 判定を2つ持たない。
  final Map<String, Object?>? visibleWhen;

  const WizardStepDefinition({
    required this.id,
    required this.title,
    this.description,
    this.layout = LayoutDefinition.single,
    this.fields = const [],
    this.visibleWhen,
  });

  /// This step as a standalone form, so the ordinary [FormDefinition] machinery
  /// (validator, normalizer, form renderer) works on one step unchanged.
  FormDefinition get form => FormDefinition(
        sections: [
          SectionDefinition(
            fields: fields,
            layout: layout,
            visibleWhen: visibleWhen,
          ),
        ],
      );

  @override
  List<Object?> get props =>
      [id, title, description, layout, fields, visibleWhen];
}

File: java/src/main/java/io/hatake/core/WizardStepDefinition.java
package io.hatake.core;

import java.util.List;
import java.util.Map;

/**
 * ステップ入力（{@code type: wizard}）の 1 ステップ。
 * <b>id と見出しを持つ {@link SectionDefinition}</b>、と思えばいい。
 *
 * <p>描画専用のキー（{@code description} / {@code layout}）は、この版の他のモデルと
 * 同じ方針で持たない（バックエンドは検証・クエリのために定義を読む）。
 *
 * @param id ステップ識別子
 * @param title ステップ見出し
 * @param fields そのステップの入力項目
 * @param visibleWhen そのステップを丸ごと出す条件（null なら常に出す）
 */
public record WizardStepDefinition(
        String id,
        String title,
        List<FieldDefinition> fields,
        Map<String, Object> visibleWhen) {

    /** 条件を持たない普通のステップ。 */
    public WizardStepDefinition(String id, String title, List<FieldDefinition> fields) {
        this(id, title, fields, null);
    }

    /**
     * このステップだけのフォーム。{@link FormValidator} にこれを渡せば、
     * <b>そのステップの項目だけ</b>を検証できる（「次へ」相当）。
     *
     * <p>条件はそのまま区画の {@code visibleWhen} になる＝<b>隠れているステップは
     * 検証もしない</b>（「隠れた区画は検証しない」がそのまま効く）。判定を2つ持たない。
     */
    public FormDefinition form() {
        return new FormDefinition(
                List.of(new SectionDefinition(title, fields, visibleWhen)));
    }
}
